File: backend/src/main/java/com/app/backend/domain/group/dto/response/GroupResponse.java
package com.app.backend.domain.group.dto.response;

import com.app.backend.domain.group.entity.Group;
import com.app.backend.domain.group.entity.GroupRole;
import com.app.backend.domain.group.entity.MembershipStatus;
import com.app.backend.global.util.AppUtil;
import java.util.List;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;

public class GroupResponse {

    public static Detail toDetail(final Group group) {
        return Detail.builder()
                     .id(group.getId())
                     .categoryName(group.getCategory().getName())
                     .name(group.getName())
                     .province(group.getProvince())
                     .city(group.getCity())
                     .town(group.getTown())
                     .description(group.getDescription())
                     .recruitStatus(group.getRecruitStatus().name())
                     .maxRecruitCount(group.getMaxRecruitCount())
                     .currentMemberCount(Math.toIntExact(
                             group.getMembers().stream().filter(m -> m.getStatus() == MembershipStatus.APPROVED
                                                                     && !m.getDisabled()).count()
                     )).createdAt(AppUtil.localDateTimeToString(group.getCreatedAt()))
                     .groupLeaders(group.getMembers().stream().filter(m -> m.getStatus() == MembershipStatus.APPROVED
                                                                           && m.getGroupRole() == GroupRole.LEADER
                                                                           && !m.getDisabled())
                                        .map(m -> m.getMember().getNickname()).toList())
                     .build();
    }

    public static Detail toDetail(final Group group,
                                  final boolean isApplying,
                                  final boolean isMember,
                                  final boolean isAdmin) {
        return Detail.builder()
                     .id(group.getId())
                     .categoryName(group.getCategory().getName())
                     .name(group.getName())
                     .province(group.getProvince())
                     .city(group.getCity())
                     .town(group.getTown())
                     .description(group.getDescription())
                     .recruitStatus(group.getRecruitStatus().name())
                     .maxRecruitCount(group.getMaxRecruitCount())
                     .currentMemberCount(Math.toIntExact(
                             group.getMembers().stream().filter(m -> m.getStatus() == MembershipStatus.APPROVED
                                                                     && !m.getDisabled()).count()
                     )).createdAt(AppUtil.localDateTimeToString(group.getCreatedAt()))
                     .groupLeaders(group.getMembers().stream().filter(m -> m.getStatus() == MembershipStatus.APPROVED
                                                                           && m.getGroupRole() == GroupRole.LEADER
                                                                           && !m.getDisabled())
                                        .map(m -> m.getMember().getNickname()).toList())
                     .isApplying(isApplying)
                     .isMember(isMember)
                     .isAdmin(isAdmin)
                     .build();
    }

    public static ListInfo toListInfo(final Group group) {
        return ListInfo.builder()
                       .id(group.getId())
                       .categoryName(group.getCategory().getName())
                       .name(group.getName())
                       .province(group.getProvince())
                       .city(group.getCity())
                       .town(group.getTown())
                       .recruitStatus(group.getRecruitStatus().name())
                       .maxRecruitCount(group.getMaxRecruitCount())
                       .currentMemberCount(Math.toIntExact(
                               group.getMembers().stream().filter(m -> m.getStatus() == MembershipStatus.APPROVED
                                                                       && !m.getDisabled()).count()
                       )).createdAt(AppUtil.localDateTimeToString(group.getCreatedAt()))
                       .groupLeaders(group.getMembers().stream().filter(m -> m.getStatus() == MembershipStatus.APPROVED
                                                                             && m.getGroupRole() == GroupRole.LEADER
                                                                             && !m.getDisabled())
                                          .map(m -> m.getMember().getNickname()).toList())
                       .build();
    }

    @Getter
    @Builder(access = AccessLevel.PRIVATE)
    @AllArgsConstructor(access = AccessLevel.PRIVATE)
    public static class Detail {
        private final Long         id;
        private final String       categoryName;
        private final String       name;
        private final String       province;
        private final String       city;
        private final String       town;
        private final String       description;
        private final String       recruitStatus;
        private final Integer      maxRecruitCount;
        private final Integer      currentMemberCount;
        private final String       createdAt;
        private final Boolean      isApplying;
        private final Boolean      isMember;
        private final Boolean      isAdmin;
        private final List<String> groupLeaders;
    }

    @Getter
    @Builder(access = AccessLevel.PRIVATE)
    @AllArgsConstructor(access = AccessLevel.PRIVATE)
    public static class ListInfo {
        private final Long         id;
        private final String       categoryName;
        private final String       name;
        private final String       province;
        private final String       city;
        private final String       town;
        private final String       recruitStatus;
        private final Integer      maxRecruitCount;
        private final Integer      currentMemberCount;
        private final String       createdAt;
        private final List<String> groupLeaders;
    }

}

File: backend/src/main/kotlin/com/app/backend/domain/group/dto/response/GroupResponse.kt
package com.app.backend.domain.group.dto.response

import com.app.backend.domain.group.entity.Group
import com.app.backend.domain.group.entity.GroupRole
import com.app.backend.domain.group.entity.MembershipStatus
import com.app.backend.global.util.AppUtil
import com.fasterxml.jackson.annotation.JsonProperty

class GroupResponse {
    companion object {
        fun toDetail(group: Group) = Detail(
            id = group.id!!,
            categoryName = group.category.name,
            name = group.name,
            province = group.province,
            city = group.city,
            town = group.town,
            description = group.description,
            recruitStatus = group.recruitStatus.name,
            maxRecruitCount = group.maxRecruitCount,
            currentMemberCount = group.members.count { it.status == MembershipStatus.APPROVED && !it.disabled },
            createdAt = AppUtil.localDateTimeToString(group.createdAt),
            groupLeaders = group.members.filter { it.status == MembershipStatus.APPROVED && it.groupRole == GroupRole.LEADER && !it.disabled }
                .map { it.member.nickname!! }
        )

        fun toDetail(group: Group, isApplying: Boolean, isMember: Boolean, isAdmin: Boolean) = Detail(
            id = group.id!!,
            categoryName = group.category.name,
            name = group.name,
            province = group.province,
            city = group.city,
            town = group.town,
            description = group.description,
            recruitStatus = group.recruitStatus.name,
            maxRecruitCount = group.maxRecruitCount,
            currentMemberCount = group.members.count { it.status == MembershipStatus.APPROVED && !it.disabled },
            createdAt = AppUtil.localDateTimeToString(group.createdAt),
            isApplying = isApplying,
            isMember = isMember,
            isAdmin = isAdmin,
            groupLeaders = group.members.filter { it.status == MembershipStatus.APPROVED && it.groupRole == GroupRole.LEADER && !it.disabled }
                .map { it.member.nickname!! }
        )

        fun toListInfo(group: Group) = ListInfo(
            id = group.id!!,
            categoryName = group.category.name,
            name = group.name,
            province = group.province,
            city = group.city,
            town = group.town,
            recruitStatus = group.recruitStatus.name,
            maxRecruitCount = group.maxRecruitCount,
            currentMemberCount = group.members.count { it.status == MembershipStatus.APPROVED && !it.disabled },
            createdAt = AppUtil.localDateTimeToString(group.createdAt),
            groupLeaders = group.members.filter { it.status == MembershipStatus.APPROVED && it.groupRole == GroupRole.LEADER && !it.disabled }
                .map { it.member.nickname!! }
        )
    }

    data class Detail(
        val id: Long,
        val categoryName: String,
        val name: String,
        val province: String,
        val city: String,
        val town: String,
        val description: String,
        val recruitStatus: String,
        val maxRecruitCount: Int,
        val currentMemberCount: Int,
        val createdAt: String,
        @get:JsonProperty("isApplying") val isApplying: Boolean? = null,
        @get:JsonProperty("isMember") val isMember: Boolean? = null,
        @get:JsonProperty("isAdmin") val isAdmin: Boolean? = null,
        val groupLeaders: List<String>
    )

    data class ListInfo(
        val id: Long,
        val categoryName: String,
        val name: String,
        val province: String,
        val city: String,
        val town: String,
        val recruitStatus: String,
        val maxRecruitCount: Int,
        val currentMemberCount: Int,
        val createdAt: String,
        val groupLeaders: List<String>
    )
}